### PR TITLE
Temporal versioning: Support selecting of entities

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Adjusts the query logic for the `Read` methods and the corresponding `EntityQueryPath` to allow querying of entities as before.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203505325130329/f) _(internal)_

## 🚫 Blocked by

- #1624 

## 🔍 What does this change?

- Replaces `EntityQueryPath::Version` with `EntityQueryPath::LowerTransactionTime` (to be removed as part of [Adjust structural queries for temporal versioning](https://app.asana.com/0/0/1203491211535116/f) (_internal_))
- Adds `EntityQueryPath::{EditionId, DecisionTime, TransactionTime}`
- Adds `Lower`, `Upper`, and `Now` functions
- Uses `Now` function as `"latest"` special case
- Adjust `Read` implementation
- Sets the correct filter for getting archived entities

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs [hunk 1]
```diff
@@ -44,56 +44,15 @@ pub enum EntityQueryPath<'q> {
...
-    /// ```
-    ///
-    /// Typically, this is a timestamp, but also `"latest"` can be specified as a parameter.
-    ///
-    /// [`EntityVersion`]: crate::identifier::knowledge::EntityVersion
-    /// [`EntityEditionId`]: crate::identifier::knowledge::EntityEditionId
-    /// [`Entity`]: crate::knowledge::Entity
-    Version,
+    // TODO: DOC: https://app.asana.com/0/0/1203505325130325/f
+    EditionId,
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs]: Thanks 🎉

File: packages/graph/hash_graph/tests/rest-test.http [hunk 2]
```diff
@@ -465,7 +465,7 @@ GET http://127.0.0.1:4000/entities/{{encoded_person_a_entity_id}}
     });
 %}
 
-#### Update Person entity
+### Update Person entity
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/tests/rest-test.http:468]: We basically had commented out the heading 🙃

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/tests/rest-test.http:468]: 😅 oops

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs [hunk 3]
```diff
@@ -134,18 +135,42 @@ impl EntityVersion {
     }
 }
 
-#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, ToSql)]
+#[postgres(transparent)]
+#[repr(transparent)]
+pub struct EntityRecordId(i64);
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:141]: I added a TODO to the corresponding column in Postgres: 0582b7834e98b47b63adafdd55c0e6ca1ba37f21

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs [hunk 4]
```diff
@@ -134,18 +135,42 @@ impl EntityVersion {
...
+    pub const fn as_i64(&self) -> i64 {
+        self.0
+    }
+}
+
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, ToSchema)]
 #[serde(rename_all = "camelCase")]
 pub struct EntityEditionId {
     base_id: EntityId,
+    #[serde(skip)]
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:159]: I wouldn't like to expose it as it's likely, that we will have breaking changes in this struct anyway. If we are happy with the struct and want to use it in the backend we can still remove the skipping. 
I'd defer it until [Adjust structural queries for temporal versioning](https://app.asana.com/0/0/1203491211535116/f) _(internal)_.

File: packages/graph/hash_graph/lib/graph/src/store/query/filter.rs [hunk 5]
```diff
@@ -534,14 +511,14 @@ impl Parameter<'_> {
             }
             (Parameter::Text(text), ParameterType::Timestamp) => {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/store/query/filter.rs:514]: I don't know if we are going to remove that, but most likely we will. I added a TODO to at least reevaluate, if we still need this after [Adjust structural queries for temporal versioning](https://app.asana.com/0/0/1203491211535116/f) _(internal)_: 5a0964b771dd0bb249a99f2790887e8a7c246565

File: packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs [hunk 6]
```diff
@@ -422,7 +386,7 @@ mod tests {
 
     #[test]
     fn deserialization() {
-        assert_eq!(deserialize(["version"]), EntityQueryPath::Version);
+        assert_eq!(deserialize(["version"]), EntityQueryPath::DecisionTimeFrom);
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs:389]: Should `"version"` be `LowerTransactionTime`?

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs [hunk 7]
```diff
@@ -134,18 +135,42 @@ impl EntityVersion {
...
 
-#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq, Serialize, Deserialize, ToSchema)]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, ToSql)]
+#[postgres(transparent)]
+#[repr(transparent)]
+pub struct EntityRecordId(i64);
+
+impl EntityRecordId {
+    #[must_use]
+    pub const fn new(id: i64) -> Self {
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:145]: I wonder if we should go with `from_i64` in this case because any other kind of random ID could be generated in its `new`? No strong feelings about this, but if we want to treat this as a random opaque Id, maybe we should assume it can be generated randomly

File: packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs [hunk 8]
```diff
@@ -41,16 +42,14 @@ impl<C: AsClient> crud::Read<Entity> for PostgresStore<C> {
...
         let owned_by_id_index = compiler.add_selection_path(&EntityQueryPath::OwnedById);
-        let entity_uuid_index = compiler.add_distinct_selection_with_ordering(
-            &EntityQueryPath::Uuid,
-            Distinctness::Distinct,
-            None,
-        );
-        let version_index = compiler.add_distinct_selection_with_ordering(
-            &EntityQueryPath::Version,
+        let entity_uuid_index = compiler.add_selection_path(&EntityQueryPath::Uuid);
+        let record_id_index = compiler.add_distinct_selection_with_ordering(
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/knowledge/entity/read.rs:46]: So happy that we aren't relying on hardcoding the indexes when adding new selections here

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:159]: I agree that we will have breaking changes in the struct, but what I mean is that we can incrementally migrate.

- Let's say we _add_ `record_id`
- Next we update the workspace to stop using `version` and start using `record_id`
- Next you remove `version`

It allows us to do it in three steps, where 3 is technically a breaking change but wouldn't require updating the workspace because we'd already moved off it.

Basically a deprecation warning.

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:159]: But I don't feel strongly about this

**Review by Alfred-Mountfield — APPROVED** [review]

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:159]: In order to use it, we also need to adjust the query token parameter and have to expose `"recordId"` as a valid field on entity queries, otherwise, it couldn't be used.
I'm fine by doing this, though, but let's keep that change out of this PR, so we have more clear bounds.

**Review by thehabbos007 — COMMENTED** [review]
I had some in-flight comments pre-merge. Here they are :)

File: packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs (hunk 6 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/knowledge/entity/query.rs:389]: This was addressed in the next PR 🙂 (The test failed)

File: packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs (hunk 7 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/shared/identifier/knowledge.rs:145]: I expect this mostly to be used by `EntityRecordId::new(row.get(record_id_index))`. I'd agree, but in order to change the generate an ID inside the function this most probably would require removing the `const` anyway. If you feel strongly, I can adjust this in a follow-up PR though.
